README.md
# AutoScroll

For users on Linux or Mac, the lack of autoscroll can be a big pain. This extension can help!

### Notes

Intentional differences with Firefox:

* AutoScroll does not scroll on `contenteditable` elements; Firefox does


CHANGELOG.md
<h1>AutoScroll Changelog</h1>
<h2 title="2020-12-17">Version 5.2</h2>
<ul>
  <li>Moved "Options" page from React to Svelte.</li>
</ul>
<h2 title="2020-12-16">Version 5.1</h2>
<ul>
  <li>Fixed scrolling on pages with smooth scroll behavior enabled.</li>
</ul>
<h2 title="2020-12-16">Version 5.0</h2>
<ul>
  <li>Reworked and cleaned up the whole project.</li>
  <li>The "Options" page is now bootstrapped by React.</li>
  <li>Optimized SVG and PNG images.</li>
</ul>
<h2 title="2018-11-21">Version 4.10</h2>
<ul>
  <li>Fixing a regression that caused it to incorrectly activate on various things. (Thanks to <a href="https://github.com/Tynach">Tynach</a>)</li>
  <li>AutoScroll now works on Windows (it will disable the built-in autoscroll).</li>
</ul>
<h2 title="2018-11-03">Version 4.9</h2>
<ul>
  <li>Fixing deprecation warning in Chrome. (Thanks to <a href="https://github.com/joepvl">joepvl</a>)</li>
  <li>Changing <code>&lt;area&gt;</code> to be treated the same as links. (Thanks to <a href="https://github.com/Fry-kun">Fry-kun</a>)</li>
  <li>Enabling scrolling on disabled and readonly text inputs. (Thanks to <a href="https://github.com/Fry-kun">Fry-kun</a>)</li>
</ul>
<h2 title="2016-07-12">Version 4.8</h2>
<ul>
  <li>Changing the scroll images to be SVG. (Thanks to <a href="https://github.com/metarmask">metarmask</a>)</li>
  <li>Improving the quality of the scroll images and the icons.</li>
</ul>
<h2 title="2016-02-03">Version 4.7</h2>
<ul>
  <li>Fixed AutoScroll not working on some sites.</li>
</ul>
<h2 title="2016-01-08">Version 4.6</h2>
<ul>
  <li>Added in partial support for scrolling SVG documents.</li>
  <li>Middle clicking on the horizontal scrollbar no longer activates autoscroll.</li>
</ul>
<h2 title="2015-10-03">Version 4.5</h2>
<ul>
  <li>Fixed middle click paste not working on YouTube comments.</li>
  <li>"Scroll by using (Ctrl/⌘ + Left Click)" is now enabled by default.</li>
</ul>
<h2 title="2015-09-28">Version 4.4</h2>
<ul>
  <li>Fixed it so that textboxes are unfocused when scrolling starts.</li>
  <li>Fixed AutoScroll not working in Chrome versions 29 to 34.</li>
</ul>
<h2 title="2015-09-27">Version 4.3</h2>
<ul>
  <li>Fixed a bug that caused scrolling to not work on some websites.</li>
</ul>
<h2 title="2015-09-26">Version 4.2</h2>
<ul>
  <li>Fixed a serious bug that caused scrolling to get "stuck".</li>
</ul>
<h2 title="2015-09-25">Version 4.1</h2>
<ul>
  <li>Fixed a bug that caused scroll to not work on certain sites.</li>
</ul>
<h2 title="2015-09-25">Version 4.0</h2>
<ul>
  <li>Removed support for extremely old versions of Chrome (version 28 or older).</li>
  <li>It should now be impossible for a site or extension to break the styles for AutoScroll.</li>
  <li>Improved the performance of scrolling on all sites.</li>
  <li>Changed the default move speed to 10, because of the improved performance.</li>
</ul>
<h2 title="2014-12-21">Version 3.9</h2>
<ul>
  <li>Fixed scrolling not working on most sites.</li>
</ul>
<h2 title="2014-04-07">Version 3.8</h2>
<ul>
  <li>Fixed scrolling not working on most sites.</li>
</ul>
<h2 title="2014-04-06">Version 3.7</h2>
<ul>
  <li>Fixed a bug where scrolling did not work on <a href="https://plus.google.com/">Google+</a>.</li>
  <li>Fixed the options page not working.</li>
</ul>
<h2 title="2014-03-16">Version 3.6</h2>
<ul>
  <li>Fixed a bug where certain sites (like <a href="http://blog.dogecoin.com/">this one</a>) would get stuck scrolling.</li>
</ul>
<h2 title="2014-02-01">Version 3.5</h2>
<ul>
  <li>Added in the ability to scroll using (Ctrl/⌘ + Left Click)</li>
  <li>Added in the ability to disable scrolling using (Middle Click)</li>
</ul>
<h2 title="2013-12-27">Version 3.4</h2>
<ul>
  <li>Fixed scrolling not working on old versions of Chrome (&lt; 24.0)</li>
</ul>
<h2 title="2013-12-25">Version 3.3</h2>
<ul>
  <li>Have to revert from document_start to document_end in order to fix <a href="https://easywebsoc.td.com/waw/idp/login.htm?execution=e1s1">this website</a> constantly refreshing.</li>
</ul>
<h2 title="2013-12-25">Version 3.2</h2>
<ul>
  <li>The scroll icon should now be completely fixed for all websites, hopefully.</li>
</ul>
<h2 title="2013-12-24">Version 3.1</h2>
<ul>
  <li>Fixed <a href="http://9to5mac.com/">http://9to5mac.com/</a> not working.</li>
  <li>Fixed some sites allowing scrolling in both directions, but the icon only showing horizontal/vertical.</li>
</ul>
<h2 title="2013-12-20">Version 3.0</h2>
<ul>
  <li>Use <a href="https://developer.mozilla.org/en/docs/Web/API/window.requestAnimationFrame">requestAnimationFrame</a> for smoother scrolling.</li>
  <li>Fixed a bug with Change Colors.</li>
  <li>Fixed a bug with <a href="http://www.willhaben.at">www.willhaben.at</a></li>
  <li>Fixed being unable to click links on <a href="http://www.wind.it/it/privati/">http://www.wind.it/it/privati/</a></li>
  <li>Fixed being unable to click links on XML pages (like <a href="http://www.opengl.org/sdk/docs/man/xhtml/glDrawElements.xml">this one</a>)</li>
  <li>Cursor should now return to normal when scroll ends.</li>
</ul>
<h2 title="2012-10-23">Version 2.9</h2>
<ul>
  <li>Fixed a bug where scrolling while a text field is focused causes Chrome to scroll to the text field.</li>
  <li>Changed the default cursor to "move" from "crosshair".</li>
</ul>
<h2 title="2012-06-01">Version 2.8</h2>
<ul>
  <li>Worked around a bug in Chrome where the icon and crosshair wouldn't show up.</li>
  <li>Fixed a bug where changing certain settings in the options page was severely broken.</li>
</ul>
<h2 title="2010-11-06">Version 2.7.5</h2>
<ul>
  <li>AutoScroll is now much stricter about how it applies styles to the cursor element. This fixes issues with extensions like <a href="https://chrome.google.com/extensions/detail/jbmkekhehjedonbhoikhhkmlapalklgn">Change Colors</a>.</li>
</ul>
<h2 title="2010-10-23">Version 2.7</h2>
<ul>
  <li>Changes made in the options page are now synced immediately. You no longer need to refresh the page!</li>

src/contentScript.js
import defaults from './defaultOptions'
import math from './helpers/math'
import { getStyleFromAngle } from './helpers/cursor'

import originBoth from './images/origin/both.svg'
import originHorizontal from './images/origin/horizontal.svg'
import originVertical from './images/origin/vertical.svg'

// TODO: Restore listener on option changes
chrome.storage.local.get(defaults, function (options) {
    //     // Update the options when they change
    //     chrome.storage.onChanged.addListener(function (o, s) {
    //         console.assert(s === 'local')
    //         for (let k in o) {
    //             let x = o[k]
    //             if ('newValue' in x) {
    //                 options[k] = x.newValue
    //             } else if ('oldValue' in x) {
    //                 options[k] = defaults[k]
    //             }
    //         }
    //     })

    function image(o) {
        if (o.width && o.height) {
            return originBoth
        } else if (o.width) {
            return originHorizontal
        } else {
            return originVertical
        }
    }

    const state = {
        timeout: null,

        oldX: null,
        oldY: null,

        dirX: 0,
        dirY: 0,

        click: false,
        scrolling: false,
    }

    let htmlNamespace = 'http://www.w3.org/1999/xhtml'

    let htmlNode = document.documentElement

    // This is needed to support SVG
    let bodyNode = document.body ? document.body : htmlNode

    // Keep default scroll behavior
    // TODO: Handle current container behavior?
    // TODO: Test for SVG
    const scrollBehavior = htmlNode.style.scrollBehavior

    // The timer that does the actual scrolling; must be very fast so that the scrolling is smooth
    function startCycle(elem, scroller, root) {
        // This is needed to support SVG
        let scrollX = root ? window.scrollX : scroller.scrollLeft,
            scrollY = root ? window.scrollY : scroller.scrollTop

        function loop() {
            state.timeout = requestAnimationFrame(loop)

            let scrollWidth = scroller.scrollWidth - elem.clientWidth,
                scrollHeight = scroller.scrollHeight - elem.clientHeight

            scrollX += state.dirX
            scrollY += state.dirY

            if (scrollX < 0) {
                scrollX = 0
            } else if (scrollX > scrollWidth) {
                scrollX = scrollWidth
            }

            if (scrollY < 0) {
                scrollY = 0
            } else if (scrollY > scrollHeight) {
                scrollY = scrollHeight
            }

            // This is needed to support SVG
            if (root) {
                // This triggers a reflow
                window.scroll(scrollX, scrollY)
            } else {
                // This triggers a reflow
                scroller.scrollLeft = scrollX
                scroller.scrollTop = scrollY
            }
        }

        loop()
    }

    function shouldSticky(x, y) {
        return (
            options.stickyScroll &&
            /*state.stickyScroll && */ math.hypot(x, y) < options.dragThreshold
        )
    }

    function scale(value) {
        return value / options.moveSpeed
    }

    // This is needed to make AutoScroll work in SVG documents
    let outer = document.createElementNS(htmlNamespace, 'auto-scroll')

    // https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow
    let shadow = outer.attachShadow
        ? outer.attachShadow({ mode: 'closed' })
        : outer.createShadowRoot
        ? outer.createShadowRoot()
        : // TODO hack for Chrome 29 to 34, remove this later
          outer.webkitCreateShadowRoot()

    // This is needed to make AutoScroll work in SVG documents
    let inner = document.createElementNS(htmlNamespace, 'div')
    // TODO hack to make it so that Chrome doesn't repaint when scrolling
    inner.style.setProperty('transform', 'translateZ(0)')
    inner.style.setProperty('display', 'none')
    inner.style.setProperty('position', 'fixed')
    inner.style.setProperty('left', '0px')
    inner.style.setProperty('top', '0px')
    inner.style.setProperty('width', '100%')
    inner.style.setProperty('height', '100%')
    inner.style.setProperty('z-index', '2147483647') // 32-bit signed int
    inner.style.setProperty('background-repeat', 'no-repeat')

    shadow.appendChild(inner)

    function mousewheel(event) {
        // TODO is this a good idea ?
        stopEvent(event, true)
    }

    function mousemove(event) {
        // TODO is this a good idea ?
        stopEvent(event, true)

        let x = event.clientX - state.oldX,
            y = event.clientY - state.oldY

        if (math.hypot(x, y) > options.moveThreshold) {
            //state.stickyScroll = false;

            inner.style.setProperty(
                'cursor',
                getStyleFromAngle(math.angle(x, y))
            )

            // 10 = 5
            // 5  = 10
            // 1  = 50
            if (options.sameSpeed) {
                x = math.max(x, 1) * 50
                //(Options.get("moveSpeed") * 0.04);
                y = math.max(y, 1) * 50
                //(Options.get("moveSpeed") * 0.04);
            }

            x = scale(x)
            y = scale(y)

            if (options.shouldCap) {
                x = math.max(x, options.capSpeed)
                y = math.max(y, options.capSpeed)
            }

            state.dirX = x
            state.dirY = y
        } else {
            normalCursor()

            state.dirX = 0
            state.dirY = 0
        }
    }

    function mouseup(event) {
        // TODO is this a good idea ?
        stopEvent(event, true)

        let x = event.clientX - state.oldX,
            y = event.clientY - state.oldY

        if (state.click || !shouldSticky(x, y)) {
            unclick()
        } else {
            state.click = true
        }
    }

    function unclick() {
        cancelAnimationFrame(state.timeout)
        state.timeout = null

        removeEventListener('wheel', mousewheel, true)
        removeEventListener('mousemove', mousemove, true)
        removeEventListener('mouseup', mouseup, true)

        normalCursor()

        inner.style.removeProperty('background-image')
        inner.style.removeProperty('background-position')
        inner.style.setProperty('display', 'none')

        state.oldX = null
        state.oldY = null

        state.dirX = 0
        state.dirY = 0

        state.click = false
        state.scrolling = false

        // Restore scroll behavior
        htmlNode.style.setProperty('scroll-behavior', scrollBehavior)
    }

    function normalCursor() {
        inner.style.removeProperty('cursor')
    }

    function show(o, x, y) {
        state.scrolling = true
        state.oldX = x
        state.oldY = y

        startCycle(o.element, o.scroller, o.root)

        addEventListener('wheel', mousewheel, true)
        addEventListener('mousemove', mousemove, true)
        addEventListener('mouseup', mouseup, true)

        inner.style.setProperty('background-image', 'url("' + image(o) + '")')
        inner.style.setProperty(
            'background-position',
            x - 13 + 'px ' + (y - 13) + 'px'
        )

        inner.style.removeProperty('display')

        // Force normal scroll behavior to fix the unresponsive movements
        htmlNode.style.setProperty('scroll-behavior', 'auto')
    }

    function isInvalid(elem) {
        return (
            elem.isContentEditable ||
            (elem.localName === 'a' && elem.href) ||
            (elem.localName === 'area' && elem.href) ||
            (elem.localName === 'textarea' && isEditableText(elem)) ||
            (elem.localName === 'input' && isEditableText(elem))
        )
    }

    function isEditableText(elem) {
        return !(elem.disabled || elem.readOnly)
    }

    function isValid(elem) {
        if (options.scrollOnLinks) {
            return true
        } else {
            while (true) {
                if (elem == null) {
                    // TODO is this correct ?
                    return false
                } else if (
                    elem === document ||
                    elem === htmlNode ||
                    elem === bodyNode
                ) {
                    return true

                    // TODO better check for this
                } else if (elem.host) {
                    elem = elem.host
                } else if (isInvalid(elem)) {
                    return false
                } else {
                    elem = elem.parentNode
                }
            }
        }
    }

    function canScroll(style) {
        return style === 'auto' || style === 'scroll'
    }

    /**
     * Shows scrollbar:
     *   <html>   <body>
     *   visible  visible
     *   visible  auto
     *   visible  scroll
     *   auto     visible
     *   auto     auto
     *   auto     scroll
     *   auto     hidden
     *   scroll   visible
     *   scroll   auto
     *   scroll   scroll
     *   scroll   hidden
     *
     * Does not show scrollbar:
     *   <html>   <body>
     *   visible  hidden
     *   hidden   visible
     *   hidden   auto
     *   hidden   scroll
     *   hidden   hidden
     */
    function canScrollTop(html, body) {
        switch (html) {
            case 'visible':
                return body !== 'hidden'

            case 'auto':
            case 'scroll':
                return true

            default:
                return false
        }
    }

    // TODO this isn't quite correct, but it's close enough
    function findScrollTop(element) {
        let scroller = document.scrollingElement
            ? document.scrollingElement
            : bodyNode

        let htmlStyle = getComputedStyle(htmlNode)
        let bodyStyle = getComputedStyle(bodyNode)

        let width =
            canScrollTop(htmlStyle.overflowX, bodyStyle.overflowX) &&
            scroller.scrollWidth > element.clientWidth

        let height =
            canScrollTop(htmlStyle.overflowY, bodyStyle.overflowY) &&
            scroller.scrollHeight > element.clientHeight

        if (width || height) {
            return {
                element: element,
                scroller: scroller,
                width: width,
                height: height,
                root: true,
            }
        } else {
            return null
        }
    }

    function findScrollNormal(elem) {
        let style = getComputedStyle(elem)

        let width =
            canScroll(style.overflowX) && elem.scrollWidth > elem.clientWidth

        let height =
            canScroll(style.overflowY) && elem.scrollHeight > elem.clientHeight

        if (width || height) {
            return {
                element: elem,
                scroller: elem,
                width: width,
                height: height,
                root: false,
            }
        } else {
            return null
        }
    }

    // TODO this should handle the case where <body> has its own scrollbar (separate from the viewport's scrollbar)
    function findScroll(elem) {
        if (options.innerScroll) {
            while (
                elem !== document &&
                elem !== htmlNode &&
                elem !== bodyNode
            ) {
                // TODO is this correct ?
                if (elem == null) {
                    return null

                    // TODO better check for this
                } else if (elem.host) {
                    elem = elem.host
                } else {
                    let x = findScrollNormal(elem)

                    if (x === null) {
                        elem = elem.parentNode
                    } else {
                        return x
                    }
                }
            }
        }

        // TODO hack needed to work around non-spec-compliant versions of Chrome
        //      https://code.google.com/p/chromium/issues/detail?id=157855
        if (document.compatMode === 'CSS1Compat') {
            return findScrollTop(htmlNode)
        } else {
            return findScrollTop(bodyNode)
        }
    }

    function stopEvent(e, preventDefault) {
        e.stopImmediatePropagation()
        e.stopPropagation()

        if (preventDefault) {
            e.preventDefault()
        }
    }

    htmlNode.appendChild(outer)

    addEventListener(
        'mousedown',
        function (e) {
            if (state.scrolling) {
                stopEvent(e, true)
            } else {
                let path = e.composedPath()
                // TODO use e.target instead of null ?
                let target = path.length === 0 ? null : path[0]

                if (
                    target != null &&
                    ((e.button === 1 && options.middleClick) ||
                        (e.button === 0 &&
                            (e.ctrlKey || e.metaKey) &&
                            options.ctrlClick)) &&
                    // Make sure the click is not on a scrollbar
                    // TODO what about using middle click on the scrollbar of a non-<html> element ?
                    e.clientX < htmlNode.clientWidth &&
                    e.clientY < htmlNode.clientHeight &&
                    isValid(target)
                ) {
                    let elem = findScroll(target)

                    if (elem !== null) {
                        stopEvent(e, true)
                        show(elem, e.clientX, e.clientY)
                    }
                }
            }
        },
        true
    )
})
